src/plots/waterfallPlot.ts

```diff
@@ -0,0 +1,219 @@
+            if (i === 0) {
+              return Math.abs(yScale.scale(Math.abs(currentValue)) - yScale.scale(0));
+            } else {
+              var priorSubtotal = this._subtotals[i - 1];
+              return Math.abs(yScale.scale(currentSubtotal) - yScale.scale(priorSubtotal));
+            }
+          }
+        };
+      }
+
+      attrToProjector["class"] = (d, i, dataset) => {
+        var baseClass = "";
+        if (this.attr("class") != null) {
+          baseClass = this.attr("class").accessor(d, i, dataset) + " ";
+        }
+        var isTotal = totalAccessor(d, i, dataset);
+        if (isTotal) {
+          return baseClass + Waterfall._BAR_TOTAL_CLASS;
+        } else {
+          var delta  = this.y().accessor(d, i, dataset);
+          return baseClass + (delta > 0 ? Waterfall._BAR_GROWTH_CLASS : Waterfall._BAR_DECLINE_CLASS);
+        }
+      };
+
+      return attrToProjector;
+    }
+
+    protected _onDatasetUpdate() {
+      this._updateSubtotals();
+      super._onDatasetUpdate();
+      return this;
+    }
+
+    private _calculateSubtotalsAndExtent(dataset: Dataset) {
+      var min = Number.MAX_VALUE;
+      var max = Number.MIN_VALUE;
+      var total = 0;
+      var startValue = -Infinity;
+      dataset.data().forEach((datum, index) => {
+        var currentValue = this.y().accessor(datum, index, dataset);
+        var isTotal = this.total().accessor(datum, index, dataset);
+        if (!isTotal || index === 0) {
+          total += currentValue;
+        }
+        this._subtotals.push(total);
+        if (total < min) {
+          min = total;
+        }
+        if (total > max) {
+          max = total;
+        }
+        if (startValue === -Infinity && isTotal) {
+          var adjustment = currentValue - total;
+          for (var i = 0; i < this._subtotals.length; i++) {
+            this._subtotals[i] += adjustment;
+          }
+          startValue = this._subtotals[i];
+          total += adjustment;
+          min += adjustment;
+          max += adjustment;
+        }
+      });
+      this._extent = [min, max];
+    }
+
+    private _drawConnectors() {
+      var attrToProjector = this._generateAttrToProjector();
+      var dataset = this.datasets()[0];
+      for (var datumIndex = 1; datumIndex < dataset.data().length; datumIndex++) {
+        var prevIndex = datumIndex - 1;
+        var datum = dataset.data()[datumIndex];
+        var prevDatum = dataset.data()[prevIndex];
+        var x = attrToProjector["x"](prevDatum, prevIndex, dataset);
+        var x2 = attrToProjector["x"](datum, datumIndex, dataset) + attrToProjector["width"](datum, datumIndex, dataset);
+        var y = attrToProjector["y"](datum, datumIndex, dataset);
+        if ((this._subtotals[datumIndex] > 0 && this._subtotals[datumIndex] > this._subtotals[prevIndex]) ||
+            (this._subtotals[datumIndex] < 0 && this._subtotals[datumIndex] >= this._subtotals[prevIndex])) {
```

Comment:
This used to be a couple of nested if statements that was eventually consolidated down to this.

- If your current sum is positive and has a greater subtotal than the prior bar, you'll want to draw from the bottom of your current bar (first case)
- If your current sum is negative and has a greater or equal subtotal relative to the prior bar, you'll also want to draw from the bottom of your current bar (second case)

The reason the second one looks dodgy is because it's essentially the inverted case to the top one (<=) but ends up getting flipped in the other direction due to operating in negative space.